src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -581,6 +628,130 @@ fn max_slice_length<'p, 'a: 'p, 'tcx: 'a, I>(
+                            Some(Interval { lo, hi: hi - offset, ty })
+                        };
+                    }
+                }
+                None
+            }
+            ConstantValue(val) => {
+                let ty = val.ty;
+                val.assert_bits(ty).map(|val| Interval { lo: val, hi: val, ty })
+            }
+            Single | Variant(_) | Slice(_) => {
+                None
+            }
+        }
+    }
+
+    fn offset_sign(ty: Ty<'tcx>, (lo, hi): (u128, u128), forwards: bool) -> (u128, u128) {
+        use syntax::ast::IntTy::*;
+        match ty.sty {
+            ty::TyInt(int_ty) => {
+                macro_rules! offset_sign_for_ty {
+                    ($ity:ident, $uty:ty) => {{
+                        let min = Wrapping($ity::MIN as $uty);
+                        if forwards {
+                            ((Wrapping(lo as $uty) + min).0 as u128,
+                             (Wrapping(hi as $uty) + min).0 as u128)
+                        } else {
+                            ((Wrapping(lo as $uty) + min).0 as $ity as u128,
+                             (Wrapping(hi as $uty) + min).0 as $ity as u128)
+                        }
+                    }}
+                }
+                match int_ty {
+                    Isize => offset_sign_for_ty!(isize, usize),
+                    I8    => offset_sign_for_ty!(i8, u8),
+                    I16   => offset_sign_for_ty!(i16, u16),
+                    I32   => offset_sign_for_ty!(i32, u32),
+                    I64   => offset_sign_for_ty!(i64, u64),
+                    I128  => offset_sign_for_ty!(i128, u128),
+                }
+            }
+            ty::TyUint(_) | ty::TyChar => {
+                (lo, hi)
+            }
+            _ => bug!("`Interval` should only contain integer types")
+        }
+    }
+
+    fn into_inner(self) -> (u128, u128) {
+        (self.lo, self.hi)
+    }
+}
+
+/// Given a pattern in a `match` and a collection of ranges corresponding to the
+/// domain of values of a type (say, an integer), return a new collection of ranges
+/// corresponding to those ranges minus the ranges covered by the pattern.
+fn ranges_subtract_pattern<'a, 'tcx>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
+                                     pat_ctor: &Constructor<'tcx>,
+                                     ranges: Vec<Constructor<'tcx>>)
+                                     -> Vec<Constructor<'tcx>> {
+    if let Some(pat_interval) = Interval::from_ctor(pat_ctor) {
+        let mut remaining_ranges = vec![];
+        let mut ranges: Vec<_> = ranges.into_iter().filter_map(|r| {
+            Interval::from_ctor(&r).map(|i| i.into_inner())
+        }).collect();
+        for (subrange_lo, subrange_hi) in ranges {
+            if pat_interval.lo > subrange_hi || pat_interval.hi < subrange_lo {
+                // The pattern doesn't intersect with the subrange at all,
+                // so the subrange remains untouched.
```

Comment:
You could construct the intersection by `i = a.start.max(b.start)..=a.end.min(b.end)`, and the two halves of the subtraction `a - b` by `sl = a.start..=i.start-1` and `sh = i.end+1..=a.end` (be careful with overflow).

AFAIK, the empty condition is `range.start > range.end` (for inclusive ranges).

Since I don't think you need the intersection *by itself*, the two halves can be:
`sl = a.start..=a.start.max(b.start)-1` and
`sh = a.end.min(b.end)+1..=a.end`
(when `a.start.max(b.start) <= a.end.min(b.end)`, otherwise `sl = sh = a`)

Therefore:
```rust
if a.start > b.end || b.start > a.end {
    // No intersection.
    remaining_ranges.push(a);
} else {
    // Overflow is now not a problem, because of the conditions.
    if b.start > a.start {
        remaining_ranges.push(a.start..=b.start-1);
    }
    if b.end < a.end {
        remaining_ranges.push(b.end+1..=a.end);
    }
}
```

---

I've just checked and your code corresponds to this, but you have the 4 possibilities of my two small `if`s expanded out - **my main suggestion** from this comment would be to orthogonalize those.